mixxc: init at 0.2.2

Daru-san opened the pull request:
## Description of changes
Add package mixxc from https://github.com/Elvyria/mixxc

Closes #305855

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

Daru-san committed:
mixxc: init at 0.2.2

Aleksanaa commented on pkgs/by-name/mi/mixxc/package.nix:
```diff
@@ -0,0 +1,56 @@
+  nativeBuildInputs = [pkg-config installShellFiles];
+
+  buildInputs = [
+    pkg-config
```
This should be in `nativeBuildInputs`

Aleksanaa commented on pkgs/by-name/mi/mixxc/package.nix:
```diff
@@ -0,0 +1,56 @@
+    gtk4.dev
+    mold
+    patchelf
+    clang
```
Also `nativeBuildInputs`, `rustPlatform` already has this?

Aleksanaa commented on pkgs/by-name/mi/mixxc/package.nix:
```diff
@@ -0,0 +1,56 @@
+
+  buildInputs = [
+    pkg-config
+    libpulseaudio.dev
```
```suggestion
    libpulseaudio
```

`.dev` is already assumed in `buildInputs`.

Aleksanaa commented on pkgs/by-name/mi/mixxc/package.nix:
```diff
@@ -0,0 +1,56 @@
+    gtk4-layer-shell.dev
+    gtk4.dev
+    mold
+    patchelf
```
You mean `autoPatchelfHook`? This should also go to `nativeBuildInputs`, or (very possibly) not needed.

Aleksanaa commented on pkgs/by-name/mi/mixxc/package.nix:
```diff
@@ -0,0 +1,56 @@
+    libpulseaudio.dev
+    gtk4-layer-shell.dev
+    gtk4.dev
+    mold
```
Does it specifically depends on this linker? We should already have a linker.

Aleksanaa reviewed (commented):
Please reformat with `nixfmt-rfc-style`, since you've not included spaces in lists

Daru-san committed:
mixxc: update build inputs

Removed build inputs that were not used in the build process
Added build inputs for `wayland` and `X11` features

Daru-san commented:
About the build inputs, how would I go about adding libraries that are runtime dependencies but not required at build time?

Aleksanaa commented on pkgs/by-name/mi/mixxc/package.nix:
```diff
@@ -0,0 +1,59 @@
+  installShellFiles,
+  enableWayland ? false,
+  enableSass ? false,
+  enableX11 ? false,
```
These can all be enabled by default.

Daru-san replied:
I realize that is a good idea now that you mention it

Aleksanaa commented on pkgs/by-name/mi/mixxc/package.nix:
```diff
@@ -0,0 +1,59 @@
+  ];
+
+  nativeBuildInputs = [
+    pkg-config
```
recommend `wrapGAppsHook4`, just you don't know when people will request it (missing icon theme?)

Aleksanaa reviewed (commented):
Please reformat with `nixfmt-rfc-style`, since you've not included spaces in lists

Aleksanaa commented on pkgs/by-name/mi/mixxc/package.nix:
```diff
@@ -0,0 +1,65 @@
+  enableSass ? true,
+  enableX11 ? true,
+}:
+with lib;
```
DON'T USE THIS!

Daru-san replied:
`with lib` should not be used there? I did not know that, sorry

Aleksanaa replied:
See https://github.com/NixOS/nixpkgs/issues/208242

Daru-san replied:
Oh, I see, is there anywhere one can see or learn of all of these little details that I've missed?
 It would be easier if I could avoid these things to begin with to prevent having dozens of commits

AndersonTorres replied:
> Oh, I see, is there anywhere one can see or learn of all of these little details that I've missed? It would be easier if I could avoid these things to begin with to prevent having dozens of commits

https://nix.dev/guides/best-practices.html

Daru-san committed:
mixxc: use `buildFeatures`, enable all features by default

Daru-san committed:
mixxc: use `wrapGAppsHook4`

Fixes icons not showing up properly

Daru-san committed:
mixxc: reformat using `nixfmt-rfc-style`

Aleksanaa reviewed (approved):
Got it working!

![image](https://github.com/NixOS/nixpkgs/assets/42209822/10c83a38-3301-45a2-bf79-67cac0c467bf)
